File: nexus/models/cycle.py
from datetime import datetime
from enum import StrEnum
from uuid import uuid4

from pydantic import BaseModel, Field


class CycleTrigger(StrEnum):
    scheduled = "scheduled"
    event = "event"
    manual = "manual"


class CycleStatus(StrEnum):
    running = "running"
    completed = "completed"
    failed = "failed"
    aborted = "aborted"


class Cycle(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    started_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: datetime | None = None
    trigger: CycleTrigger = CycleTrigger.scheduled
    status: CycleStatus = CycleStatus.running
    observe_at: datetime | None = None
    detect_at: datetime | None = None
    diagnose_at: datetime | None = None
    generate_at: datetime | None = None
    validate_at: datetime | None = None
    apply_at: datetime | None = None
    verify_at: datetime | None = None
    incidents_detected: int = 0
    fixes_applied: int = 0
    errors: list[str] = Field(default_factory=list)
    target_id: str = ""

File: internal/model/cycle.go
package model

import "time"

// Cycle records one full run of the closed loop against a target.
type Cycle struct {
	ID          string       `json:"id" db:"id"`
	StartedAt   time.Time    `json:"started_at" db:"started_at"`
	CompletedAt *time.Time   `json:"completed_at,omitempty" db:"completed_at"`
	Trigger     CycleTrigger `json:"trigger" db:"trigger"`
	Status      CycleStatus  `json:"status" db:"status"`

	// Phase timestamps
	ObserveAt  *time.Time `json:"observe_at,omitempty" db:"observe_at"`
	DetectAt   *time.Time `json:"detect_at,omitempty" db:"detect_at"`
	DiagnoseAt *time.Time `json:"diagnose_at,omitempty" db:"diagnose_at"`
	GenerateAt *time.Time `json:"generate_at,omitempty" db:"generate_at"`
	ValidateAt *time.Time `json:"validate_at,omitempty" db:"validate_at"`
	ApplyAt    *time.Time `json:"apply_at,omitempty" db:"apply_at"`
	VerifyAt   *time.Time `json:"verify_at,omitempty" db:"verify_at"`

	// Results
	IncidentsDetected int      `json:"incidents_detected" db:"incidents_detected"`
	FixesApplied      int      `json:"fixes_applied" db:"fixes_applied"`
	Errors            []string `json:"errors,omitempty" db:"errors"`

	TargetID string `json:"target_id" db:"target_id"`
}

// CycleTrigger records what started a cycle.
type CycleTrigger string

const (
	TriggerScheduled CycleTrigger = "scheduled"
	TriggerEvent     CycleTrigger = "event"
	TriggerManual    CycleTrigger = "manual"
)

// CycleStatus reflects the overall state of a cycle run.
type CycleStatus string

const (
	CycleRunning   CycleStatus = "running"
	CycleCompleted CycleStatus = "completed"
	CycleFailed    CycleStatus = "failed"
	CycleAborted   CycleStatus = "aborted"
)
